apps/os/backend/durable-objects/estate-builds.ts

```diff
@@ -0,0 +1,392 @@
+
+  // Periodic alarm to detect stale builds
+  async alarm() {
+    const now = Date.now();
+    const heartbeats = await this.ctx.storage.get<Record<string, number>>("heartbeats");
+    if (!heartbeats) {
+      await this.scheduleNextAlarm();
+      return;
+    }
+
+    let changed = false;
+    for (const [buildId, last] of Object.entries(heartbeats)) {
+      if (now - last > 2 * 60_000) {
+        // 2 minutes without heartbeat → failed
+        try {
+          await this.appendLog(buildId, now, "stdout", "[BUILD FAILED]");
+          await this.broadcast({
+            type: "LOG",
+            buildId,
+            stream: "stdout",
+            message: "[BUILD FAILED]",
+            ts: now,
+          });
+        } catch (_err) {
+          void 0;
+        }
+        await this.broadcast({ type: "STATUS", buildId, status: "failed", ts: now });
+        // Update DB and invalidate org queries
+        const estateId = (await this.ctx.storage.get<string>(`estate_for_build:${buildId}`)) || "";
+        if (estateId) {
+          await this._updateBuildStatus(estateId, buildId, "failed");
+          await this.ctx.storage.put(`status:${buildId}`, "failed");
+        }
+        delete heartbeats[buildId];
+        changed = true;
+      }
+    }
+    if (changed) await this.ctx.storage.put("heartbeats", heartbeats);
+    await this.scheduleNextAlarm();
+  }
+
+  private async handleGetLogs(request: Request): Promise<Response> {
+    const url = new URL(request.url);
+    const buildId = url.searchParams.get("buildId");
+    if (!buildId) return new Response("Missing buildId", { status: 400 });
+
+    const rows = await this.ctx.storage.sql.exec(
+      "SELECT ts, stream, message FROM logs WHERE build_id = ? ORDER BY ts ASC",
+      buildId,
+    );
+    const logs: Array<{ ts: number; stream: string; message: string }> = [];
+    for (const r of rows) {
+      logs.push({ ts: Number(r.ts), stream: String(r.stream), message: String(r.message) });
+    }
+    return new Response(JSON.stringify({ buildId, logs }), {
+      headers: { "Content-Type": "application/json" },
+    });
+  }
+
+  private async ensureInitialized(): Promise<void> {
+    const initialized = (await this.ctx.storage.get("__init__")) as boolean | undefined;
+    if (!initialized) {
+      await this.ctx.storage.sql.exec(
+        "CREATE TABLE IF NOT EXISTS logs (build_id TEXT, ts INTEGER, stream TEXT, message TEXT)",
+      );
+      // Helpful index for replay and retention
+      await this.ctx.storage.sql.exec(
+        "CREATE INDEX IF NOT EXISTS idx_logs_build_ts ON logs (build_id, ts)",
+      );
+      await this.ctx.storage.put("__init__", true);
+      await this.scheduleNextAlarm();
+    }
+    // Purge logs older than 90 days on boot-up
+    const retentionBefore = Date.now() - 90 * 24 * 60 * 60 * 1000;
+    try {
+      await this.ctx.storage.sql.exec("DELETE FROM logs WHERE ts < ?", retentionBefore);
+    } catch (_err) {
+      // best-effort
+    }
```

Comment:
what's the benefit of this? do we expect builds to last this long? if they do, do we really want to not see the errors?